### PR TITLE
Module 2 Project Exercise 

### PR DESCRIPTION
**Pull request by JoeWFisher**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -115,3 +115,4 @@ env/
 ### Configuration files ###
 .env
 .vscode/settings.json
+config.py
```
Comment (Jonesey13) [inline .gitignore:118]: Setup instructions should probably be mentioned in the README file

File: app.py [hunk 2]
```diff
@@ -1,31 +1,29 @@
 from flask import Flask, render_template, request, redirect, url_for
 import session_items as session
```
Comment (Jonesey13) [inline app.py]: This file is no longer used (along with `flask_config.py` and the SECRET_KEY variable)

This also breaks the setup as `python-dotenv` has been removed from the dependency list

File: app.py [hunk 3]
```diff
@@ -1,31 +1,29 @@
...
+import trello as trello
 
 app = Flask(__name__)
 app.config.from_object('flask_config.Config')
```
Comment (Jonesey13) [inline app.py]: line can be removed

**Comment by Jonesey13** [comment]
`requirements.txt` needs to be removed and re-added in git as it's currently being tracked as a binary file (`git rm --cached` is your friend here)

File: trello.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
+from config import KEY, TOKEN, BoardID, TTD, DO
```
Comment (Jonesey13) [inline trello.py]: BoardId is unused here.

File: trello.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        ('key', KEY),
+        ('token', TOKEN),
+        ('fields', 'name,idList')
+    )
+
+    r = requests.get('https://api.trello.com/1/boards/5eef26d3f3a754437122f88b/cards', params=params)
+    data = r.json()
+    item_list = []
+    for card in data:
+        if card['idList'] == TTD:
```
Comment (Jonesey13) [inline trello.py]: `TTD` & `DO` are not terribly descriptive names

File: trello.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+
+def fetch_all_items():
+    params = (
+        ('key', KEY),
+        ('token', TOKEN),
+        ('fields', 'name,idList')
+    )
+
+    r = requests.get('https://api.trello.com/1/boards/5eef26d3f3a754437122f88b/cards', params=params)
```
Comment (Jonesey13) [inline trello.py]: The BoardId should be used here

File: trello.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+def create_new_item(name):
+    params = (
+        ('key', KEY),
+        ('token', TOKEN),
+        ('name', name),
+        ('idList', TTD)
+    )
+
+    requests.post('https://api.trello.com/1/cards', params=params)
```
Comment (Jonesey13) [inline trello.py:36]: BaseUrl should be extracted out to a constant

File: trello.py (hunk 4 above)
Comment (Jonesey13) [reply trello.py]: Also `TTD` and `DO` lists could be inferred from their names

File: trello.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
+from config import KEY, TOKEN, BoardID, TTD, DO
+import requests 
+import json
+
+class Item:
```
Comment (Jonesey13) [inline trello.py]: `Item` class should probably be moved to its own file

File: trello.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    for card in data:
+        if card['idList'] == TTD:
+            card['idList'] = 'Not Started'
+        elif card['idList'] == DO:
+            card['idList'] = 'Complete'
+
+        if [item for item in item_list if item.id == card['id']]:
+            pass
+        else:
+            item_list.append(Item(id=card['id'], status=card['idList'], title=card['name']))
```
Comment (Jonesey13) [inline trello.py]: Might be worth extracting to a conversion `@classmethod` on the `Item` class (called `from_trello_card` or something similar)

**Review by Jonesey13 — CHANGES_REQUESTED** [review]
Nearly there! Everything works except that the setup is broken because of dependency you no longer need (`python-dotenv`).

Once this is fixed please ping for re-review and I'll approve.

**Comment by Jonesey13** [comment]
Do we need all the dependencies in `requirements.txt`?

**Comment by JoeWFisher** [comment]
I'm not sure if I've sorted the setup dependency issue

**Comment by Jonesey13** [comment]
@JoeWFisher Please could you re-submit this for review on Aptem?

**Comment by Jonesey13** [comment]
> I'm not sure if I've sorted the setup dependency issue

Yes the application is still breaking due to the code in `flask_config.py`:
```python
    SECRET_KEY = os.environ.get('SECRET_KEY')
    if not SECRET_KEY:
        raise ValueError("No SECRET_KEY set for Flask application. Did you forget to run setup.sh?")
```

The environment variables in `.env` only get setup when running `flask run` if `python-dotenv` is an installed dependency. 

However you don't actually use SECRET_KEY anymore (as you're no longer using cookies so don't need the key to encrypt them), so instead you can just remove this file and the line in app.py that uses it.

File: .gitignore (hunk 1 above)
Comment (hugh-emerson) [reply .gitignore:118]: You still need to update the README. At the moment you are expecting developers to work out for themselves that config.py is required and what it should look like. 

Slightly nicer than an ignored config.py file would be to use environment variables. Add them to the .env.template and your .env file, and then access them in code with os.environ.get, so all your python can be committed. 

Either way, you need to make sure that someone can clone and run the application by just following the README.

**Comment by hugh-emerson** [comment]
The module 2 changes themselves look fine but two general points about your project:

You have poetry files since module 4, so you should delete requirements.txt, requirements-frozen.txt, setup.sh

Update your README.md file! It should include:
- Requirements to run the app (Trello account and credentials)
- How to run the app (poetry install & poetry run flask run)
- Requirements to run tests (Firefox, geckodriver)
- How to run tests (poetry run pytest)
- The fact that you can run the app in vagrant

**Pull request merged**